fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1763,29 +1794,123 @@ private void deleteRecordsWhereCheckIndexes() {
+            final Collection<RecordType> recordTypesForIndex = recordMetaData.recordTypesForIndex(index);
+
+            if (Key.Expressions.hasRecordTypePrefix(index.getRootExpression())) {
+                // If the index has a record type prefix, we need to reject the deletion, as the referenced
+                // record type key is different for the synthetic type and the requested type
+                return false;
+            }
+            String constituentName = null;
+            for (RecordType indexRecordType : recordTypesForIndex) {
+                final SyntheticRecordType<?> syntheticRecordType = (SyntheticRecordType<?>)indexRecordType;
+                if (syntheticRecordType instanceof UnnestedRecordType) {
+                    UnnestedRecordType unnestedRecordType = (UnnestedRecordType) syntheticRecordType;
+                    UnnestedRecordType.NestedConstituent parent = unnestedRecordType.getParentConstituent();
+                    if (recordType != null && !recordType.equals(parent.getRecordType())) {
+                        // If the delete is limited to a single type, only process indexes on unnested records of
+                        // precisely that type. In theory, we could use the constituent name as a predicate here,
+                        // but if the different record types have different constituent names, then they can't
+                        // be used to form a prefix.
+                        return false;
+                    }
+                    if (constituentName != null && !constituentName.equals(parent.getName())) {
+                        return false;
+                    }
+                    constituentName = parent.getName();
+                } else {
+                    // JoinedRecordTypes are difficult to handle correctly, for a few reasons. First,
+                    // if we don't have a recordType comparison, then we have to match the comparison
+                    // to _all_ constituents. That may actually be possible if the relevant fields are
+                    // used in the join conditions, but it requires additional matching. Additionally,
+                    // we have to check whether any of the Joins are outer joins, as if they are, then
+                    // we either need to make sure that _all_ join constituents are deleted or reject
+                    // the deletion, as the correct operation would replace all deleted records with
+                    // nulls
+                    return false;
+                }
+            }
+            if (constituentName == null) {
+                return false;
+            }
+            if (recordType != null) {
+                // In the first loop, we have validated that there is a single constituent that contains elements
+                // of the record type we're looking to delete. In this loop, make sure that there isn't a constituent
+                // with that name that is of a different type
```

Comment:
Hm, so, this loop is here because in an earlier version of the code, the first loop would just ignore any unnested records when `recordType != null` where the record type didn't match the desired type. So this loop was there to validate that none of those ignored constituents had the same name as the one we were going to add to the predicate, because that would have been incorrect.

Later, though, I realized that there was a possibility that someone might do something like:

1. Define two unnested record types on different stored types with different names for the parent constituent
2. Construct a multi-type index on those two types
3. Do a delete where on one of the stored types with no other predicates

In that case, there was a bug that would result in the entire index being cleared out. So I made the first loop also validate that the type of each parent matches `recordType` (if not null), so this loop is extraneous now. There was already a test, `deleteWhereOnMultiTypeFailsWithAmbiguousParent`, that validated this, and that test passes without this loop.

It's true, though, that we will have to be a little more careful with joined record types. I'll go ahead and remove this loop for now, though, as it's not really doing anything